README.md
# ChatGPT Zulip Bot

![GitHub License](https://img.shields.io/github/license/yeahjack/chatgpt_zulip_bot) [![Python Tests](https://github.com/yeahjack/chatgpt_zulip_bot/actions/workflows/ci.yml/badge.svg)](https://github.com/yeahjack/chatgpt_zulip_bot/actions/workflows/ci.yml) ![visitors](https://visitor-badge.glitch.me/badge?page_id=yeahjack.chatgpt_zulip_bot&left_color=green&right_color=blue)

The ChatGPT Zulip Bot is a bot that responds to users by using the ChatGPT language model. It can be used in any Zulip chat streams or private messages.

# Installation

1. Clone the repository:

```bash
git clone https://github.com/yeahjack/chatgpt_zulip_bot.git
```

2. Install the required dependencies (You might need to create a virtual env if you like):

```bash
pip install -r requirements.txt
```

3. Set up your Zulip bot:

- Go to your Zulip organization settings, and navigate to the "Your bots" section.
- Click the "Add a new bot" button and follow the prompts to create a new bot.
- Download the configuartion file and move it to this folder.
- Rename `config.ini.example` to `config.ini` and fill in the values, note that you do not have to add commas (`" "`) around the values.
Here are the explanations of parameters in `config.ini`.

| Parameters | Details |
| --- | --- |
| OPENAI_API_KEY | Your OpenAI API Key. Could be set via [this](https://platform.openai.com/account/api-keys).
| ZULIP_CONFIG | The filename of your zulip bot configuration file.
| USER_ID | The ID of the admin. Used for notifications.
| BOT_ID | The ID of the zulip bot. Used for positioning.
| API_VERSION | The version of OpenAI models. e.g. `gpt-3.5-turbo` |

4. Start the bot:

```bash
python chatgpt_zulip_bot.py
```

## Proxies

If you need to set a proxy for the OpenAI API, the normal method would be
```python
import os
os.environ["http_proxy"] = "http://127.0.0.1:7890"
os.environ["https_proxy"] = "http://127.0.0.1:7890"
```
Remember to replace both strings to your proxy address and port.

# Usage

To use the bot, mention it in any Zulip stream or send it a private message. You can mention the bot by typing `@bot_name` where the name is what you gave to the bot when you created it.
Except normal texts, the bot also accepts the following commands

## Commands
* `/help`: print this usage information.
* `/end`: end the current conversation. Bot will answer questions based on the context of the conversation. If a conversation reaches its 3000 token limit (you will see a message: "ERROR: OpenAI API rate limit exceeded. Please retry."), then you must restart the conversation with `/end`.

# Testing

If you forked this repo and did some changes, you can run the tests to make sure everything is working fine. Each time you push a commit, the tests will be automatically run by GitHub Actions. Note that you need to set `API_VERSION` and `OPENAI_API_KEY` in GitHub secrets.

# Contributing and bug reports

Feel free to leave an [issue](https://github.com/yeahjack/chatgpt_zulip_bot/issues) if you have any questions or suggestions.
Pull requests are also welcomed. If you are interested in contributing this project, a good place to start is the `chatgpt_zulip_bot.py` and `chatgpt.py` files. You can customize the bot's behavior by modifying it.


NEWS.md
# Updates
## April 2, 2023
- Use logging instead of printing.

## April 2, 2023
- Fixed bugs during the refactoring.
- Improve the pytest support.

## April 1, 2023
- Refactored the codes by [@GiggleLiu](https://github.com/GiggleLiu) to obey the DRY principle. Thanks for his contribution!
- Initial pytest support added.

## March 31, 2023
- Improved academic prompts added to the bot, type `/help` to access them.
- Context feature now disabled when using academic prompts.
- Fixed issue with conversation trimming when single prompt exceeding token limit.

## March 28, 2023
- Add `/end` to end the current conversation, which is shorter and fits Zulip syntexes.

## March 27, 2023
- Upgraded the model from `text-davinci-003` to `gpt-3.5-turbo` by default. It can be configured in `config.ini`.
- Added contextual support, allowing the bot to answer questions based on the context of the conversation. Type `停止会话` or `end the conversation` to end the current conversation.
- The output will also show numbers of tokens used in the conversation.
- Implemented conversation history trimming to ensure it stays within OpenAI's maximum token limit.
- Fixed a bug causing the bot to crash during long conversations.
- Resolved an issue where the bot would reply privately when mentioned in a group conversation.




test_all.py
import pytest
from chatgpt import OpenAI
from configparser import ConfigParser
import tiktoken
import os

api_version = os.environ["API_VERSION"]
api_key = os.environ["OPENAI_API_KEY"]


@pytest.fixture
# Test for chatgpt.py
def openai_object():
    return OpenAI(api_version=api_version, api_key=api_key)


def test_trim_conversation_history(openai_object):
    def count_tokens(message, api_version):
        if "gpt-3.5-turbo" in api_version:
            encoding = tiktoken.encoding_for_model("gpt-3.5-turbo")
        elif "gpt-4" in api_version:
            encoding = tiktoken.encoding_for_model("gpt-4")
        else:
            raise ValueError("Unsupported API version, your API version is: " + api_version)
        return len(encoding.encode(message))

    history = [
        "User: Hello",
        "AI: Hi there!",
        "User: How are you?",
        "AI: I'm doing great! How about you?",
        "User: I'm fine, thank you."
    ]

    # Calculate token count for each message
    message_tokens = [count_tokens(
        msg, openai_object.api_version) for msg in history]

    # Set max_tokens to the sum of tokens of the last two messages
    max_tokens = message_tokens[-1] + message_tokens[-2]

    trimmed_history = openai_object.trim_conversation_history(
        history, max_tokens)
    assert len(
        trimmed_history) == 2, "Trimmed history should have only the last two messages"


def test_chatgpt_response_strings(openai_object):
    prompts = [
        "/1",
        "/end",
        "hi",
        "/polish_en this is a test cases",
        "/polish_zh 测试用例这一个是。",
        "/find_grammar_mistakes this is a test cases",
        "/zh-en 这是一个测试用例",
        "/en-zh This is a test case",
        "/en-ac This is a test case",
        "/ex_code_zh print('This is a test case.')"
    ]

    try:
        for prompt in prompts:
            result = openai_object.get_chatgpt_response(0, prompt)
            assert isinstance(
                result, str), f"Result should be a string for prompt: {prompt}"
    finally:
        # Clear the conversation history for test cases.
        openai_object.user_conversations[0] = []


chatgpt_zulip_bot.py
# chat_gpt_zulip_bot.py

import zulip
import re
from chatgpt import OpenAI
import atexit
from configparser import ConfigParser
import logging


class ChatGPTZulipBot(zulip.Client):
    def __init__(self, config_file, user_id, bot_id, ai):
        super().__init__(config_file=config_file)
        self.user_id = user_id
        self.bot_id = bot_id
        # the OpenAI instance
        self.ai = ai

    def send_notification(self, message):
        self.send_message(
            {
                "type": "private",
                "to": [self.user_id],
                "content": message,
            }
        )

    def process_message(self, msg):
        sender_email = msg["sender_email"]
        message_content = msg["content"]
        message_type = msg["type"]
        if msg["sender_id"] != self.bot_id:
            if message_content.startswith("@**ChatGPT**"):
                stream_id = msg.get("stream_id", None)
                topic = msg.get("subject", None)
                prompt = re.sub("@\*\*ChatGPT\*\*", "",
                                message_content).strip()
                response = self.ai.get_chatgpt_response(
                    msg["sender_email"], prompt)
                self.send_message(
                    {
                        "type": "stream",
                        "to": stream_id,
                        "subject": topic,
                        "content": response,
                    }
                )

            if message_type == "private":
                prompt = message_content
                response = self.ai.get_chatgpt_response(
                    msg["sender_email"], prompt)
                self.send_message(
                    {
                        "type": "private",
                        "to": sender_email,
                        "content": response,
                    }
                )


def on_exit(bot):
    bot.send_notification("NOTICE: The ChatGPT bot is now offline.")


def serve(configfile):
    config = ConfigParser()
    config.read(configfile)

    OPENAI_API_KEY = config["settings"]["OPENAI_API_KEY"]
    OPENAI_API_VERSION = config["settings"]["API_VERSION"]

    ZULIP_CONFIG = config["settings"]["ZULIP_CONFIG"]
    USER_ID = int(config["settings"]["USER_ID"])
    BOT_ID = int(config["settings"]["BOT_ID"])

    ai = OpenAI(OPENAI_API_VERSION, OPENAI_API_KEY)
    bot = ChatGPTZulipBot(ZULIP_CONFIG, USER_ID, BOT_ID, ai)
    bot.send_notification("NOTICE: The ChatGPT bot is now online.")
    print("Successfully started the ChatGPT bot.")

    atexit.register(on_exit, bot)

    bot.call_on_each_message(bot.process_message)


if __name__ == "__main__":
    logging.basicConfig(filename="bot.log")
    serve("config.ini")


chatgpt.py
# chatgpt.py
import openai
import tiktoken
import logging

prompt_table = {
    "/polish_en": "Below is a paragraph from an academic paper. Polish the writing to meet the academic style, \
improve the spelling, grammar, clarity, concision and overall readability. When neccessary, rewrite the whole sentence. \
Furthermore, list all modification and explain the reasons to do so.",
    "/polish_zh": "作为一名中文学术论文写作改进助理，你的任务是改进所提供文本的拼写、语法、清晰、简洁和整体可读性，同时分解长句，减少重复，并提供改进建议。请只提供文本的更正版本，避免包括解释。请编辑以下文本：",
    "/find_grammar_mistakes": "Below is a paragraph from an academic paper. Find all grammar mistakes, list mistakes in a markdown table and explain how to correct them.",
    "/zh-en": "As a translator, your task is to accurately translate text from Chinese to English. \
Please pay attention to context and accurately explain phrases and proverbs. \
Below is the text you need to translate: ",
    "/en-zh": "As a translator, your task is to accurately translate text from English to Chinese. \
Please pay attention to context and accurately explain phrases and proverbs. \
Below is the text you need to translate: ",
    "/en_ac": "Please translate following sentence to English with academic writing, and provide some related authoritative examples: ",
    "/ex_code_zh": "请解释以下代码："
}

help_str = """# Usage

To use the bot, mention it in any Zulip stream or send it a private message. You can mention the bot by typing `@bot_name` where the name is what you gave to the bot when you created it.
Except normal texts, the bot also accepts the following commands

## Commands
* `/help`: print this usage information.
* `/end`: end the current and start a new conversation. Bot will answer questions based on the context of the conversation. If see a rate limit exceed error after approximately 3500 token limit is reached in a single conversation, then you must restart the conversation with `/end`.
""" + "\n".join(["* `%s`: (Prompt) %s" % (k, v) for k, v in prompt_table.items()])


class OpenAI(object):
    def __init__(self, api_version, api_key, max_content_length=4097-300):
        # setup API version and API keys
        self.api_version = api_version
        self.api_key = api_key
        openai.api_key = api_key
        # Maintain a dictionary to store conversation history per user
        self.user_conversations = {}
        # maximum number of tokens in a single conversation
        self.max_content_length = max_content_length

    def trim_conversation_history(self, history, max_tokens):
        # Determine the appropriate encoding based on the API version.
        # Ref: https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
        if "gpt-3.5-turbo" in self.api_version:
            encoding = tiktoken.encoding_for_model("gpt-3.5-turbo")
        elif "gpt-4" in self.api_version:
            encoding = tiktoken.encoding_for_model("gpt-4")
        elif 'text-embedding-ada' in self.api_version:
            encoding = tiktoken.encoding_for_model('text-embedding-ada-002')
        elif "text-davinci-002" in self.api_key:
            encoding = tiktoken.encoding_for_model("text-davinci-002")
        elif 'text-davinci-003' in self.api_key:
            encoding = tiktoken.encoding_for_model("text-davinci-003")
        else:
            return "OpenAI API Version Wrong!"

        tokens = 0
        trimmed_history = []
        for message in reversed(history):
            # Get the number of tokens for the current message
            message_tokens = len(encoding.encode(message))
            if tokens + message_tokens <= max_tokens:
                trimmed_history.insert(0, message)
                tokens += message_tokens
            else:
                break
        if tokens > 0:
            return trimmed_history
        else:
            return "Cannot Trim Conversation History! Please start a new conversation."

    def get_chatgpt_response(self, user_id, prompt):
        global help_str
        if user_id not in self.user_conversations:
            self.user_conversations[
                user_id
            ] = []  # Create a new conversation history for a new user

        # Check if user input is "停止会话" or "end the conversation"
        if prompt == "停止会话" or prompt.lower() == "end the conversation" or prompt.lower() == "/end":
            # Clear the conversation history for the user
            self.user_conversations[user_id] = []
            return "The conversation has been ended and the context has been cleared."
        elif prompt == "/help":
            return help_str
        else:
            prompt, ret_code = prompt_manager(prompt)
            conversation_history = self.user_conversations[user_id]
            if ret_code == 1: 
                # Contexual conversation
                conversation_history.append(
                    f"User: {prompt}"
                )  # Add user input to conversation history
            elif ret_code == 2:  # command not found
                return prompt
        logging.info(f"API call from user: {user_id}")
        while True:
            messages = [
                {
                    "role": "system",
                    "content": "You are an AI language model trained to assist with a variety of tasks.",
                }
            ]  # System message for context
            if ret_code == 0:
                messages.append({'role': 'user', 'content': prompt})
            else:
                for message in conversation_history:
                    role, content = message.split(": ", 1)
                    messages.append({"role": role.lower(), "content": content})
            try:
                response = openai.ChatCompletion.create(
                    model=self.api_version,
                    messages=messages,
                    max_tokens=1200,
                    temperature=0.5,
                )

                if response.choices:
                    role = response["choices"][0]["message"]["role"]
                    reply = (
                        response["choices"][0]["message"]["content"].strip().replace(
                            "", "")
                    )
                    conversation_history.append(
                        f"{role}: {reply}"
                    )  # Add AI response to conversation history
                    return (
                        reply
                        + "\n------\nTokens used: "
                        + str(response.usage.prompt_tokens) + " (prompt) + "
                        + str(response.usage.completion_tokens) + " (answer)"
                        + " = " + str(response.usage.total_tokens) + "/" + str(self.max_content_length)
                    )
                else:
                    err_msg = "Sorry, I couldn't generate a response."
                    logging.error(err_msg)
                    return err_msg

            except openai.error.RateLimitError:
                err_msg = "Sorry, OpenAI API rate limit exceeded. Please end the conversation by typing `/end` or retry."
                logging.error(err_msg)
                return err_msg

            except openai.error.OpenAIError as e:
                if "Please reduce the length" in str(e):
                    conversation_history = self.trim_conversation_history(
                        conversation_history, self.max_content_length
                    )
                else:
                    logging.error(f"An error occurred on line {e.__traceback__.tb_lineno}: {e}")
                    return f"An error occurred."

            except Exception as e:
                logging.error(f"An error occurred on line {e.__traceback__.tb_lineno}: {e}")
                return f"An error occurred."

# take a message as input, returns a prompt message and a return code
# return code 0: command message: no context
# return code 1: normal contextual message
# return code 2: command not found
def prompt_manager(message):
    # Academic prompts which might be helpful.
    # Credits to https://github.com/binary-husky/chatgpt_academic/blob/b1e33b0f7aa9e69061d813262eb36ac297d49d0d/functional.py
    if message.startswith("/"):
        # get command name and message
        result = message.split(" ", 1)
        if len(result) > 1:  # has space
            name, msg = result
        else:
            name, msg = result[0], ""
        # get prompt and return code from name
        if name in prompt_table:
            prompt = prompt_table[name]
            logging.info(f"Command hit: {name}")
            return prompt + "\n\n" + msg, 0
        else:
            msg = f"Sorry, command not found: `{name}`, type `/help` to get the list of commands."
            logging.error(msg)
            return msg, 2
    else:
        return message, 1
